hips/hip-9999.md

```diff
@@ -0,0 +1,317 @@
+- **Renderer plugins:**
+Maybe more than one in a single chart? based on mime type, eg .yaml .pkl, etc
+ie, gotemplate, CUE, pkl, etc.
+- **Ordering/sequencing plugins:**
+The ability to specify the order of applying resources to the kubeapi. eg, depends_on between resources
+- **Waiting/status plugins:**
+For example, kstatus could be the default. Another plugin could add more robust options like waiting for a custom thing to happen for a particular resource before another resource is applied (we realize this is not ideal but someone may need this for legacy applications, and could create a plugin for this)
+- **Post-renderer plugins:**
+Post-render chart output YAML
+- **Storage backend plugins:**
+Where to store Helm release metadata. Currently defaults to k8s secrets, but also could be configmap (helm v2 compatibility), and SQL (Postgres). Could also support OCI Artifacts, etc.
+
+## How to teach this
+
+<!-- How to teach users, new and experienced, how to apply the HIP to their work. -->
+
+WIP
+
+- Documentation that very clear to follow
+- Blog posts and conference talks explaining how to take advantage of the new plugin system
+- Example plugin templates
+
+## Reference implementation
+
+<!-- Link to any existing implementation and details about its state, e.g. proof-of-concept. -->
+
+We will be using Wasm for Plugin Interfacing, leveraging the Extism project. See <https://extism.org/> and specifically the Extism [Go SDK](https://github.com/extism/go-sdk) and [Go PDK](https://github.com/extism/go-pdk).
+
+We can also use the new XTP for bindings / codegen / host functions https://github.com/extism/go-pdk?tab=readme-ov-file#generating-bindings
+    - Note we would still have responsibility to—through the host function—get the pointer. Then Extism makes it easy to readbytes, allocate memory etc. Ultimately, it makes it easy for a plugin implementer to write, but some work in the host function. And the folks at Dylibso are very happy to help us at that stage.
+    
+### Proof-of-concept
+    
+Two example Wasm plugins leveraging Extism have been prototyped. The first is an example downloader plugin, and the second is an example render plugin:
+- https://github.com/gjenkins8/helm-plugin-ocigetter
+- https://github.com/gjenkins8/helm-plugin-gotemplate-renderer/
+
+The host functions are prototyped using Go testing, in `/testdriver/main_test.go` in each Git repo. These are fairly basic to start, but proves out that this works. Creating the proper interfaces will be part of the work implementing this HIP (see "[Implementation Plan](#implementation-plan)").
+
+### Technical reasoning for Wasm plugins
+
+For a plugin system to work, Helm must be able to “invoke” the plugin with a suitable message so that the plugin can perform its operation and return a result. The traditional "[foreign function interface](https://en.wikipedia.org/wiki/Foreign_function_interface)” problem
+(source: https://kerkour.com/webassembly-wasi-preview2)
```

Comment:
This is the image this source note refers to, which helps to explain Wasm plugin invocation (exported and imported functions). Adding as a comment now for discussion. If this feels useful, we can make a version of this diagram in mermaid to include in the HIP.

<img width="730" alt="Screenshot 2025-02-18 at 12 51 37 PM" src="https://github.com/user-attachments/assets/fb46af9b-4649-45e9-8849-6dfb198e4c7a" />

Reply:
Made a better version in 76a0ea48257847c49b8ada963f49a0acfe5b6615

<img width="580" alt="Screenshot 2025-02-19 at 5 24 29 PM" src="https://github.com/user-attachments/assets/ae59f3f2-d35a-4e91-a7cc-24772c990976" />